File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.java
import java.util.*;

class Solution {
    
    static boolean visited[];
    static List<Integer> possible = new ArrayList<>();
    static int[][] edges;
    static int[] info;

    public int solution(int[] info, int[][] edges) {
        Solution.edges = edges;
        Solution.info = info;
        visited = new boolean[info.length];
        
        visited[0] = true;
        if (info[0] == 0) {
            backTrack(1, 0);
        } else {
            backTrack(0, 1);
        }
        
        return Collections.max(possible);
    }
    
    public void backTrack(int sheepCnt, int wolfCnt) {
        if (sheepCnt > wolfCnt) {
            possible.add(sheepCnt);
        } else {
            return;
        }
        
        for (int[] edge: edges) {
            int s = edge[0];
            int e = edge[1];
            if (visited[s] && !visited[e]) {
                visited[e] = true;
                if (info[e] == 0) {
                    backTrack(sheepCnt + 1, wolfCnt);
                } else {
                    backTrack(sheepCnt, wolfCnt + 1);
                }
                visited[e] = false;
            }
        }
        
        
        
    }
}

File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.py
def solution(info, edges):
    visited = [False] * len(info)
    
    cnt_wolf = 0
    cnt_sheep = 1 #문제에 항상 시작인 1이라고 나와있음
    
    possible = list()
    
    def backtrack(cnt_sheep, cnt_wolf):
        if cnt_sheep > cnt_wolf:
            possible.append(cnt_sheep)
        else:
            return
        
        for s, e in edges:
            if visited[s] and not visited[e]:
                visited[e] = True
                if not info[e]:
                    backtrack(cnt_sheep + 1, cnt_wolf)
                else:
                    backtrack(cnt_sheep, cnt_wolf + 1)
                visited[e] = False
    
    visited[0] = True
    backtrack(cnt_sheep, cnt_wolf)
    
    return max(possible)